### DISCUSSION TITLE
Is there a way to tell Kestrel not to return 411 error when Content-Length header is not set for requests with empty body?

### DISCUSSION THREAD

**Discussion by Dzoge** [dotnet/aspnetcore#28565] [Q&A] [answered]
Hey
I ran into an issue with some clients calling my REST sevice that do not send Content-Length header when the body is empty and Transfer-Encoding is not chunked. When this happens Kestrel returns **HTTP 411 Length Required** response.

I was looking for a way to intercept the request and manually inject Content-Length header with 0 but could not find the way to do it.

The server is running without a reverse-proxy in front of it and the requests go directly to Kestrel docker pod.

I searched the Kestrel source code and found a check for that specifically:
[https://github.com/dotnet/aspnetcore/blob/857bb0f416b746c9eb7b9a412443b39c1701df6b/src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs#L182](https://github.com/dotnet/aspnetcore/blob/857bb0f416b746c9eb7b9a412443b39c1701df6b/src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs#L182) 
but I could not find any way to override this behavour.

Is there any way to configure or intercept this behaviour and manually add a Content-Length header with 0 value or somehow make it pass this validation?

There is another way to solve this issue by putting a reverse-proxy in front of Kestrel that will fix the headers, but I don't want to introduce a new component in the chain if it is other way.

**Comment by davidfowl** [comment, marked as answer]
No there's no way to inject logic at this stage, it's too early. Only a reverse proxy can help you here.

cc @halter73 @Tratcher

**Comment by Tratcher** [comment]
https://tools.ietf.org/html/rfc7230#section-3.3.3
>    6.  If this is a request message and none of the above are true, then
       the message body length is zero (no message body is present).
...
>    A server MAY reject a request that contains a message body but not a
   Content-Length by responding with 411 (Length Required).

https://tools.ietf.org/html/rfc7231#section-4.3.3 (POST) doesn't say anything about 411s or requiring a non-zero body.

I think Kestrel is wrong here, it should default to no-body when there is no content-length or chunked for HTTP/1.x. I'll open an issue. (https://github.com/dotnet/aspnetcore/issues/29398)

This is a non-issue for HTTP/2 since the introduction of the END_OF_STREAM flag.

**Reply by halter73** [reply]
@Dzoge What's the value of the Transfer-Encoding header?

**Reply by Dzoge** [reply]
Transfer-Encoding header is not present at all
